### PR TITLE
More transform updates

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR addresses a few more transform changes/issues.

The primary change is that transforms are now stateless (i.e. they no longer carry their own parameters).   Stateful transforms make it very easy to accidentally introduce old and/or irrelevant parameters into a graph, and are a source for some extremely confusing and difficult bugs.  That's why this change was made.

Now, transforms only take a "parameter extraction function" that, when applied to a random variable, will extract the required transform parameters.

In other words, transform objects are no longer random variable _instance_-specific, but random variable _class_-specfic.

### CODE REVIEW

**Comment by ricardoV94** [comment]
Hopefully this will prove safer than the previous approach. 

Did you mean to pull into v4 instead of master?

Should we add a test that would fail with the old decoupled system? This v3 bug for the previous prior predictive is related to the same issue: https://github.com/pymc-devs/pymc3/issues/4484
